CMP-3149: Handle profile deprecation

yuumasato opened the pull request:
The Content Profiles can now be set as deprecated.
The Compliance Operator needs to be able to identify them and warn users of the need to make adjustments.

- When Parsing Content, it annotates deprecated profiles with `compliance.openshift.io/profile-status: deprecated`
- Whenever a `ComplianceScan` is started using a deprecated profile we add a warning to the object.
  - - From the `ComplianceScanSpec`, it find the `ProfileBundle` that matches the `ContentImage` and `ContentFile`. 
  Once a matching `ProfileBundle` is found it simply grabs the profile whose ID matches the `ComplianceScanSpec` and checks for its deprecated annotation.

yuumasato commented on pkg/controller/compliancescan/compliancescan_controller.go:
```diff
@@ -284,6 +285,30 @@ func (r *ReconcileComplianceScan) validate(instance *compv1alpha1.ComplianceScan
+	err := r.Client.Get(context.TODO(), key, profile)
+	if err != nil {
+		logger.Error(err, "Cannot get profile", profileName)
+		return reconcile.Result{}, err
```
Mock profiles added.

rhmdnd reviewed (changes requested):
We might need to either move where we perform the profile bundle look up to get this working with the existing e2e tests, or refactor the e2e tests to use `ScanSettingBindings` instead of `ComplianceScans` directly.

yuumasato commented on pkg/controller/compliancescan/compliancescan_controller.go:
```diff
@@ -282,8 +283,39 @@ func (r *ReconcileComplianceScan) validate(instance *compv1alpha1.ComplianceScan
 }
 
+func (r *ReconcileComplianceScan) notifyUseOfDeprecatedProfile(instance *compv1alpha1.ComplianceScan, logger logr.Logger) error {
+	profileBundleName := instance.Annotations[compv1alpha1.ComplianceScanProfileBundleAnnotation]
```
With the CO upgrade process in mind, would it make sense to skip profile deprecation check if `Compliancescan` doesn't have a  `compv1alpha1.ComplianceScanProfileBundleAnnotation` annotation?

I think this would skip the profile deprecation check in the current e2e tests.
But we better add other e2e tests ensure it is working though, :joy:

yuumasato replied:
I have moved away from using annotations to set the ProfileBundle in the scans.
The ComplianceScan controller now derives the ProfileBundle from the ComplianceScan name, similar to what is done in:
https://github.com/ComplianceAsCode/compliance-operator/blob/74ea72fcdd2c9ea92e4e9bd37ce58d1963581e8d/pkg/controller/compliancesuite/compliancesuite_controller.go#L436

But that means that the names of the ComplianceScans in the e2e tests need to be more meaningful, :cry:

yuumasato replied:
I have pivoted an approach that doesn't rely on the ComplianceScan name.

From the `ComplianceScanSpec`, I find the `ProfileBundle` that matches the `ContentImage` and `ContentFile`.
Once a matching `ProfileBundle` is found it simply grabs the profile whose ID matches the `ComplianceScanSpec`.

yuumasato commented on pkg/controller/compliancescan/compliancescan_controller.go:
```diff
@@ -282,8 +283,39 @@ func (r *ReconcileComplianceScan) validate(instance *compv1alpha1.ComplianceScan
+		logger.Info("ComplianceScan uses deprecated profile", instance.Name, profileName)
+		r.Recorder.Eventf(
+			instance, corev1.EventTypeWarning, "DeprecatedProfile",
+			"Profile %s is deprecated and will be removed in a future version of Compliance Operator. "+
```
Unfortunately there is no good way to convey the CO version in which a profile will be deprecated in the content.
And it actually doesn't feel right to specifically mention in the content the CO version in which a profile will be removed, after all, the CO build of the image container and release process is in control of that.

The profile deprecation will surely be in the release notes, and that definitely is the right place to mention the CO version in which a profile will be removed. Additionaly, the warning could refer folks to release notes.

rhmdnd replied:
Ack - thanks for walking through this.

xiaojiey commented:
/open

xiaojiey commented:
/reopen

yuumasato commented on pkg/controller/compliancesuite/compliancesuite_controller.go:
```diff
@@ -249,6 +250,13 @@ func (r *ReconcileComplianceSuite) reconcileScans(suite *compv1alpha1.Compliance
+		// Annotate scans using deprecated profiles
+		// We annotate during reconcile so that existing scans may also be annotated
+		if err := r.annotateScansUsingDeprecatedProfile(suite, scan, scanWrap, logger); err != nil {
+			logger.Error(err, "Could not check whether ComplianceSuite's scan uses deprecated profile", suite.Name, scan.Name, scanWrap.Profile)
```
True, but also there are e2e tests where the SuiteController cannot find the Profile.
`
{"level":"error","ts":"2025-03-03T16:35:49.459Z","logger":"suitectrl","msg":"Could not check whether ComplianceSuite's scan uses deprecated profile","Request.Namespace":"osdk-e2e-20642d4f-2eb8-45cd-9f17-c54ff8e01dc8","Request.Name":"test-suite-two-scans","test-suite-two-scans":"test-suite-two-scans-workers-scan","error":"Could not find profile","stacktrace":"github.com/ComplianceAsCode/compliance-operator/pkg/controller/compliancesuite.(*ReconcileComplianceSuite).reconcileScans\n\tgithub.com/ComplianceAsCode/compliance-operator/pkg/controller/compliancesuite/compliancesuite_controller.go:256\ngithub.com/ComplianceAsCode/compliance-operator/pkg/controller/compliancesuite.(*ReconcileComplianceSuite).Reconcile\n\tgithub.com/ComplianceAsCode/compliance-operator/pkg/controller/compliancesuite/compliancesuite_controller.go:163\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Reconcile\n\tsigs.k8s.io/controller-runtime@v0.20.2/pkg/internal/controller/controller.go:118\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.20.2/pkg/internal/controller/controller.go:328\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.20.2/pkg/internal/controller/controller.go:288\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.20.2/pkg/internal/controller/controller.go:249"}
`

Vincent056 commented:
```
{"level":"error","ts":"2025-03-05T17:44:44.207Z","logger":"scanctrl","msg":"Could not check whether the Profile used by ComplianceScan is deprecated","Request.Namespace":"osdk-e2e-7b158ce5-a22b-40e9-847c-e4195f914d70","Request.Name":"audit-profile-set-test","ComplianceScan":"audit-profile-set-test","error":"Profile.compliance.openshift.io \"\" not found","stacktrace":"github.com/ComplianceAsCode/compliance-operator/pkg/controller/compliancescan.(*ReconcileComplianceScan).phasePendingHandler\n\tgithub.com/ComplianceAsCode/compliance-operator/pkg/controller/compliancescan/compliancescan_controller.go:359\ngithub.com/ComplianceAsCode/compliance-operator/pkg/controller/compliancescan.(*ReconcileComplianceScan).Reconcile\n\tgithub.com/ComplianceAsCode/compliance-operator/pkg/controller/compliancescan/compliancescan_controller.go:191\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Reconcile\n\tsigs.k8s.io/controller-runtime@v0.20.2/pkg/internal/controller/controller.go:118\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.20.2/pkg/internal/controller/controller.go:328\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.20.2/pkg/internal/controller/controller.go:288\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.20.2/pkg/internal/controller/controller.go:249"}
```
seen this error in the operator log

rhmdnd commented on tests/e2e/framework/common.go:
```diff
@@ -176,6 +176,27 @@ func (f *Framework) PrintROSADebugInfo(t *testing.T) {
+	// suite.
+	if err := f.Client.Create(context.TODO(), origPb, nil); err != nil {
+		return "", err
+	}
```
We could wire up the defer here, too:

```go
defer f.Client.Delete(context.TODO(), origPb)
```

Looks like we're doing that each time we setup a profile bundle anyway.

yuumasato replied:
I thought of that too, but was uncertain how much it would be deferred.
Is it until we exit the function? Or exit the test?

If it is deferred until end of function, won't it be deleted right after we exit `setUpProfileBundle()?

yuumasato replied:
I think I cannot put the `defer` in the function:
```
Defer

A defer statement defers the execution of a function until the surrounding function returns. 
```

rhmdnd replied:
Ack - thank for checking.

yuumasato commented on pkg/controller/compliancescan/compliancescan_controller.go:
```diff
@@ -282,8 +283,95 @@ func (r *ReconcileComplianceScan) validate(instance *compv1alpha1.ComplianceScan
+			return nil
+		}
+		// The extends field references a profile by its full name
+		profileName = tp.Spec.Extends
```
Those are two different cases.

I think most common case is a `TailoredProfile` extending a Profile that is/got deprecated.
The other case is the user himself adding the deprecated profile annotation to their own `TailoredProfiles`

rhmdnd replied:
Ok - but we're supporting both of those now?

yuumasato replied:
Yes.

yuumasato commented on pkg/controller/compliancescan/compliancescan_controller.go:
```diff
@@ -282,8 +283,95 @@ func (r *ReconcileComplianceScan) validate(instance *compv1alpha1.ComplianceScan
+		}
+		if pbName == "" {
+			err := goerrors.New("Could not find ProfileBundle used by scan")
+			logger.Error(err, "ComplianceScan uses non-existing ProfileBundle", "ComplianceScan", instance.Name, "Profile", instance.Spec.Profile)
```
When user creates a ScanSettingBinding, it should not happen.
But it can happen when creating ComplianceScans due to typos in the `ComplianceScanSpec`, for example: https://github.com/ComplianceAsCode/compliance-operator/blob/12ef8b9e912375b72cae22dbe440368453f8a600/tests/e2e/parallel/main_test.go#L1148

yuumasato replied:
Typo fixed.

Vincent056 commented on tests/e2e/framework/common.go:
```diff
@@ -449,6 +471,41 @@ func (f *Framework) ensureTestNamespaceExists() error {
+	polledScan := &compv1alpha1.ComplianceScan{}
+
+	// Wait for profile deprecation warning event
+	err := wait.Poll(framework.RetryInterval, framework.Timeout, func() (bool, error) {
```
could you remove framework, I think you can use `RetryInterval` directly

Vincent056 replied:
also same with `Timeout`

rhmdnd commented on tests/e2e/parallel/main_test.go:
```diff
@@ -731,6 +671,242 @@ func TestSingleScanTimestamps(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	if testScan.Status.ErrorMessage != "Could not check whether the Profile used by ComplianceScan is deprecated" {
```
This is effectively testing improved validation experience, instead of waiting for the scan to error out in a later phase because the `oscap` binary can't find the profile, right?

yuumasato replied:
Yes, the scan fails much earlier when a scan uses a non-existent profile.

rhmdnd commented on tests/e2e/parallel/main_test.go:
```diff
@@ -731,6 +671,242 @@ func TestSingleScanTimestamps(t *testing.T) {
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      tpName,
+			Namespace: f.OperatorNamespace,
+			Annotations: map[string]string{
```
Ok - intersting, so this will let cluster administrators signal to users (e.g., other cluster admins?) that a tailored profile is deprecated.

yuumasato replied:
Correct, cluster admins have one more tool to manage `TailoredProfiles`

rhmdnd commented on tests/e2e/parallel/main_test.go:
```diff
@@ -948,6 +1127,30 @@ func TestSingleTailoredScanSucceeds(t *testing.T) {
 	f := framework.Global
 	scanName := framework.GetObjNameFromTest(t)
+
+	tpName := "test-tailoredprofile"
```
Interesting - did the test pass previously because we were creating the ConfigMap manually? Do we need to do that anymore?

yuumasato replied:
I think we still need to set the `TailoringConfigmap` manually.

This is usually done by the ScanSettingBinding controller, but as the test is creating the Scan directly, it needs to set the CM itself:
https://github.com/ComplianceAsCode/compliance-operator/blob/b7de0ed19163cef0397c1253f7a9248b016be33d/pkg/controller/scansettingbinding/scansettingbinding_controller.go#L544

Previously the scan was succeeding because the ComplianceScan Controller did not care whether the `TailoredProfile` existed or not. But now, it needs to exist so its deprecation annotation can be checked.

yuumasato replied:
I think this means that I need to add a `TailoringConfigMap` to some of my tests....

yuumasato replied:
@Vincent056 Tryingto reuse the `TailoringConfigMap` from the `TailoredProfile` is leading to an infinite reconcile loop. I can't figure out why.
```
{"level":"info","ts":"2025-03-13T14:09:35.724Z","logger":"scanctrl","msg":"Phase: Launching","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated"}
{"level":"info","ts":"2025-03-13T14:09:35.724Z","logger":"scanctrl","msg":"Creating PVC for scan","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated","PersistentVolumeClaim.Name":"test-scan-tailored-profile-is-deprecated","PersistentVolumeClaim.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83"}
{"level":"info","ts":"2025-03-13T14:09:35.731Z","logger":"scanctrl","msg":"Creating scan result server pod","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated"}
{"level":"info","ts":"2025-03-13T14:09:35.748Z","logger":"scanctrl","msg":"ResultServer Deployment launched","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated","Deployment.Name":"test-scan-tailored-profile-is-deprecated-rs"}
{"level":"info","ts":"2025-03-13T14:09:35.764Z","logger":"scanctrl","msg":"ResultServer Service launched","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated","Service.Name":"test-scan-tailored-profile-is-deprecated-rs"}
{"level":"info","ts":"2025-03-13T14:09:35.765Z","logger":"scanctrl","msg":"Creating a pod for node","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated","Pod.Name":"ip-10-0-46-249.us-west-1.compute.internal"}
{"level":"info","ts":"2025-03-13T14:09:35.765Z","logger":"scanctrl","msg":"Content image","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated","image":"ghcr.io/complianceascode/k8scontent:latest"}
{"level":"info","ts":"2025-03-13T14:09:35.765Z","logger":"scanctrl","msg":"Reconciling Tailoring ConfigMap","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated","ConfigMap.Name":"test-tailored-profile-is-deprecated-tp","ConfigMap.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83"}
{"level":"info","ts":"2025-03-13T14:09:40.079Z","logger":"scanctrl","msg":"Reconciling ComplianceScan","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-scan-succeeds"}
{"level":"info","ts":"2025-03-13T14:09:40.079Z","logger":"scanctrl","msg":"Phase: Launching","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-scan-succeeds"}
{"level":"info","ts":"2025-03-13T14:09:40.080Z","logger":"scanctrl","msg":"Creating PVC for scan","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-scan-succeeds","PersistentVolumeClaim.Name":"test-single-tailored-scan-succeeds","PersistentVolumeClaim.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83"}
{"level":"info","ts":"2025-03-13T14:09:40.088Z","logger":"scanctrl","msg":"Creating scan result server pod","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-scan-succeeds"}
{"level":"info","ts":"2025-03-13T14:09:40.103Z","logger":"scanctrl","msg":"ResultServer Deployment launched","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-scan-succeeds","Deployment.Name":"test-single-tailored-scan-succeeds-rs"}
{"level":"info","ts":"2025-03-13T14:09:40.121Z","logger":"scanctrl","msg":"ResultServer Service launched","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-scan-succeeds","Service.Name":"test-single-tailored-scan-succeeds-rs"}
{"level":"info","ts":"2025-03-13T14:09:40.121Z","logger":"scanctrl","msg":"Creating a pod for node","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-scan-succeeds","Pod.Name":"ip-10-0-46-249.us-west-1.compute.internal"}
{"level":"info","ts":"2025-03-13T14:09:40.121Z","logger":"scanctrl","msg":"Content image","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-scan-succeeds","image":"ghcr.io/complianceascode/k8scontent:latest"}
{"level":"info","ts":"2025-03-13T14:09:40.121Z","logger":"scanctrl","msg":"Reconciling Tailoring ConfigMap","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-scan-succeeds","ConfigMap.Name":"test-tailoredprofile-tp","ConfigMap.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83"}
{"level":"info","ts":"2025-03-13T14:09:45.689Z","logger":"scanctrl","msg":"Reconciling ComplianceScan","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-platform-scan-succeeds"}
{"level":"info","ts":"2025-03-13T14:09:45.689Z","logger":"scanctrl","msg":"Phase: Launching","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-platform-scan-succeeds"}
{"level":"info","ts":"2025-03-13T14:09:45.690Z","logger":"scanctrl","msg":"Creating PVC for scan","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-platform-scan-succeeds","PersistentVolumeClaim.Name":"test-single-tailored-platform-scan-succeeds","PersistentVolumeClaim.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83"}
{"level":"info","ts":"2025-03-13T14:09:45.701Z","logger":"scanctrl","msg":"Creating scan result server pod","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-platform-scan-succeeds"}
{"level":"info","ts":"2025-03-13T14:09:45.716Z","logger":"scanctrl","msg":"ResultServer Deployment launched","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-platform-scan-succeeds","Deployment.Name":"test-single-tailored-platform-scan-succeeds-rs"}
{"level":"info","ts":"2025-03-13T14:09:45.738Z","logger":"scanctrl","msg":"ResultServer Service launched","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-platform-scan-succeeds","Service.Name":"test-single-tailored-platform-scan-succeeds-rs"}
{"level":"info","ts":"2025-03-13T14:09:45.738Z","logger":"scanctrl","msg":"Creating a Platform scan pod","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-platform-scan-succeeds"}
{"level":"info","ts":"2025-03-13T14:09:45.738Z","logger":"scanctrl","msg":"Content image","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-platform-scan-succeeds","image":"ghcr.io/complianceascode/k8scontent:latest"}
{"level":"info","ts":"2025-03-13T14:09:45.738Z","logger":"scanctrl","msg":"Reconciling Tailoring ConfigMap","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-single-tailored-platform-scan-succeeds","ConfigMap.Name":"test-tailoredplatformprofile-tp","ConfigMap.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83"}
{"level":"info","ts":"2025-03-13T14:09:45.766Z","logger":"scanctrl","msg":"Reconciling ComplianceScan","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated"}
{"level":"info","ts":"2025-03-13T14:09:45.766Z","logger":"scanctrl","msg":"Phase: Launching","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated"}
{"level":"info","ts":"2025-03-13T14:09:45.766Z","logger":"scanctrl","msg":"Creating PVC for scan","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated","PersistentVolumeClaim.Name":"test-scan-tailored-profile-is-deprecated","PersistentVolumeClaim.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83"}
{"level":"info","ts":"2025-03-13T14:09:45.773Z","logger":"scanctrl","msg":"Creating scan result server pod","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated"}
{"level":"info","ts":"2025-03-13T14:09:45.788Z","logger":"scanctrl","msg":"ResultServer Deployment launched","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated","Deployment.Name":"test-scan-tailored-profile-is-deprecated-rs"}
{"level":"info","ts":"2025-03-13T14:09:45.805Z","logger":"scanctrl","msg":"ResultServer Service launched","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated","Service.Name":"test-scan-tailored-profile-is-deprecated-rs"}
{"level":"info","ts":"2025-03-13T14:09:45.805Z","logger":"scanctrl","msg":"Creating a pod for node","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated","Pod.Name":"ip-10-0-46-249.us-west-1.compute.internal"}
{"level":"info","ts":"2025-03-13T14:09:45.805Z","logger":"scanctrl","msg":"Content image","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated","image":"ghcr.io/complianceascode/k8scontent:latest"}
{"level":"info","ts":"2025-03-13T14:09:45.806Z","logger":"scanctrl","msg":"Reconciling Tailoring ConfigMap","Request.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83","Request.Name":"test-scan-tailored-profile-is-deprecated","ConfigMap.Name":"test-tailored-profile-is-deprecated-tp","ConfigMap.Namespace":"osdk-e2e-3d40c8af-8dff-4023-8cf4-d17fa1944c83"}
```

Vincent056 replied:
yes, that looks like the case. I wonder if it makes sense to use SSB to create the scan instead

rhmdnd reviewed (commented):
Thank you for all the e2e tests.

yuumasato commented:
/retest

yuumasato commented:
/test e2e-aws-parallel
/test e2e-aws-serial

yuumasato commented:
/test images

yuumasato commented:
/retest-required

yuumasato commented:
/retest

yuumasato commented:
/retest

rhmdnd commented:
/retest

Vincent056 commented on tests/e2e/parallel/main_test.go:
```diff
@@ -731,6 +671,292 @@ func TestSingleScanTimestamps(t *testing.T) {
+	defer f.Client.Delete(context.TODO(), testScan)
+
+	// The profile deprecation warning is sent out during Pending phase
+	if err = f.WaitForScanStatus(f.OperatorNamespace, scanName, compv1alpha1.PhasePending); err != nil {
```
I think we might miss the `PhasePending` during the retry windows

Vincent056 commented on tests/e2e/parallel/main_test.go:
```diff
@@ -731,6 +671,292 @@ func TestSingleScanTimestamps(t *testing.T) {
+	}
+	defer f.Client.Delete(context.TODO(), tp)
+
+	tailoringCM := &corev1.ConfigMap{
```
I wonder if we cloud reused the tailored profile `configmap` created by the operator

yuumasato replied:
I think we can, :)

I actually copied this CM from a cluster I had, XD.

Vincent056 commented on tests/e2e/parallel/main_test.go:
```diff
@@ -731,6 +671,292 @@ func TestSingleScanTimestamps(t *testing.T) {
+	defer f.Client.Delete(context.TODO(), testScan)
+
+	// The profile deprecation warning is sent out during Pending phase
+	if err = f.WaitForScanStatus(f.OperatorNamespace, scanName, compv1alpha1.PhasePending); err != nil {
```
same with last comment, I think we might not always run check in PhasePending

Vincent056 commented on tests/e2e/parallel/main_test.go:
```diff
@@ -731,6 +671,292 @@ func TestSingleScanTimestamps(t *testing.T) {
+	t.Parallel()
+	f := framework.Global
+
+	pbName := framework.GetObjNameFromTest(t)
```
would we need to create the profile bundle here if not using the SSB to create the scan

yuumasato commented:
I addressed some of the issues causing test failures.
But some of the failures on TailoredProfiles are misterious to me....
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/ComplianceAsCode_compliance-operator/690/pull-ci-ComplianceAsCode-compliance-operator-master-e2e-aws-parallel/1899830648651124736

yuumasato commented:
/retest

rhmdnd commented:
Looks like the e2e tests are failing on a missing event for TailoredProfile cases:

```
common.go:481: compliancescans.compliance.openshift.io "test-tailored-profile-is-deprecated" not found
common.go:517: No ComplianceScan event for profile "test-tailored-profile-is-deprecated" found
--- FAIL: TestScanTailoredProfileIsDeprecated (1800.47s) 
```

Vincent056 commented:
> Looks like the e2e tests are failing on a missing event for TailoredProfile cases:
> 
> ```
> common.go:481: compliancescans.compliance.openshift.io "test-tailored-profile-is-deprecated" not found
> common.go:517: No ComplianceScan event for profile "test-tailored-profile-is-deprecated" found
> --- FAIL: TestScanTailoredProfileIsDeprecated (1800.47s) 
> ```

I have seen one record of it in https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/ComplianceAsCode_compliance-operator/690/pull-ci-ComplianceAsCode-compliance-operator-master-e2e-aws-serial/1900238988149526528/artifacts/e2e-aws-serial/gather-extra/artifacts/events.json：

```
{
            "apiVersion": "v1",
            "count": 3,
            "eventTime": null,
            "firstTimestamp": "2025-03-13T20:09:05Z",
            "involvedObject": {
                "apiVersion": "compliance.openshift.io/v1alpha1",
                "kind": "TailoredProfile",
                "name": "test-tailored-profile-extends-deprecated",
                "namespace": "osdk-e2e-ba643d90-03f8-4658-a0e2-520788ea4fab",
                "resourceVersion": "90832",
                "uid": "75d7fb6b-0d78-4338-9fa3-43dce5c4e3e5"
            },
            "kind": "Event",
            "lastTimestamp": "2025-03-13T20:09:19Z",
            "message": "TailoredProfile test-tailored-profile-extends-deprecated is extending a deprecated Profile (ocp4-cis-1-4) that will be removed in a future version of Compliance Operator. Please consider using a newer version of this profile",
            "metadata": {
                "creationTimestamp": "2025-03-13T20:09:05Z",
                "name": "test-tailored-profile-extends-deprecated.182c75adce3b792a",
                "namespace": "osdk-e2e-ba643d90-03f8-4658-a0e2-520788ea4fab",
                "resourceVersion": "90929",
                "uid": "a1f5a2b5-c939-49bd-8b3f-e0be6613800f"
            },
            "reason": "DeprecatedTailoredProfile",
            "reportingComponent": "tailoredprofile-controller",
            "reportingInstance": "",
            "source": {
                "component": "tailoredprofile-controller"
            },
            "type": "Warning"
        },
```

yuumasato commented:
> Looks like the e2e tests are failing on a missing event for TailoredProfile cases:
> 
> ```
> common.go:481: compliancescans.compliance.openshift.io "test-tailored-profile-is-deprecated" not found
> common.go:517: No ComplianceScan event for profile "test-tailored-profile-is-deprecated" found
> --- FAIL: TestScanTailoredProfileIsDeprecated (1800.47s) 
> ```

Found the issue. The `TailoredProfile` was missing the product in the rule name, :joy:...
Should be fixed now.

Vincent056 commented:
/lgtm
